File: 2024/2B.cpp
#include <bits/stdc++.h>
#include "readfile.hpp"

int get_problem(const std::vector<int> &numbers) {
	bool increasing = false, decreasing = false;
	int prev = -1;
	for (int i = 0; i < numbers.size(); ++i) {
		if (prev == -1) {
			prev = numbers[i];
			continue;
		}
		if (prev < numbers[i])
			increasing = true;
		else if (prev > numbers[i])
			decreasing = true;
		if (std::abs(prev - numbers[i]) > 3 || prev == numbers[i] || increasing && decreasing || !increasing && !decreasing)
			return i;
		prev = numbers[i];
	}
	return -1;
}


int main(int argc, char const *argv[])
{
	int res = 0;
	std::vector<std::string> file = readfile("2.in");
	if (file.empty())
		return 1;

	for (const auto &line : file) {
		std::istringstream iss(line);
		std::vector<int> lv;
		int n;
		while (iss >> n) {
			lv.push_back(n);
		}
		if (get_problem(lv) == -1) {
			++res;
			continue;
		}

		// not safe / strike 1
		// try removing the problem number
		for (int i = 0; i < lv.size(); ++i) {
			std::vector<int> lv2;
			for (int j = 0; j < lv.size(); ++j) {
				if (j == i)
					continue;
				lv2.push_back(lv[j]);
			}
			if (get_problem(lv2) == -1) {
				++res;
				break;
			}
		}
	}

	// print
	std::cout << res << std::endl;
}

File: 2024/2B.py
res = 0


vals = []
with open('2.txt', 'r') as file:
    for line in file:
        parts = line.strip().split(" ")
        inlist = []
        for num in parts:
            inlist.append(int(num))
        vals.append(inlist)


slists = []


for val in vals:
    print(val)  
    isafe = False  

    for i in range(len(val)): 
        row = val.copy()  
        row.pop(i)  
        print(row)  
        

        increasing = False
        decreasing = False
        safe = True
        prev = -1
        
        for j in row:
            if prev == -1:
                prev = j
                continue
            if prev < j:
                increasing = True
            elif prev > j:
                decreasing = True
            if (abs(prev - j) > 3 or prev == j or 
                (j > prev and decreasing) or (j < prev and increasing) or 
                (increasing and decreasing) or (not increasing and not decreasing)):
                safe = False
                break 
            prev = j
        
        if safe:
            isafe = True  
            break  

    slists.append(isafe) 

res = sum(slists)

print(res)
